grails-app/controllers/org/pih/warehouse/JsonController.groovy

```diff
@@ -1738,25 +1740,33 @@ class JsonController {
     def productChanged = {
         Product product = Product.get(params.productId)
         Organization supplier = Organization.get(params.supplierId)
+        Organization destinationParty = Organization.get(params.destinationPartyId)
         List productSuppliers = []
         if (product && supplier) {
             productSuppliers = ProductSupplier.findAllByProductAndSupplier(product, supplier)
         }
-        productSuppliers = productSuppliers.collect {[
-            id: it.id,
-            code: it.code,
-            supplierCode: it.supplierCode,
-            text: it.code + ' ' + it.name,
-            manufacturerCode: it.manufacturerCode,
-            manufacturer: it.manufacturer?.id,
-        ]}
+        productSuppliers = productSuppliers.collect {
+            ProductSupplierPreference preference = it.productSupplierPreferences.find { it.destinationParty == destinationParty }
```

Comment:
Is there any reason not to move this into the query above. If not, ProductSupplier probably needs its own service at this point. So let's create ProductSupplierService and add a method to retrieve ProductSuppliers using the given criteria.

Reply:
Which query do you mean? What we do above is checking if preference for this destination party exists and if it exists then we check if validation code is different than hide (and then we include this product source in select) or if preference for this destination party doesn't exist at all (and then we also include it). I'm not really sure if I know how would you like it to look like.

Reply:
Yeah I guess I was wondering if we could make it work so you can match on preferences in the query.
```
def productSuppliers = ProductSupplier.createCriteria().list { 
  eq("product", product)
  eq("supplier", suppler)
  productSupplierPreferences { 
    and { 
       eq("destinationParty", destinationParty)
       not { 
          eq("validationCode", ValidationCode.HIDE)
       }
    }
}
```
}

Reply:
You might need to include something like this 
```
(productSupplierPreferences is empty or ...)
```